src/Analysis.cpp

```diff
@@ -279,13 +281,48 @@ void doAliasReasoning(Instruction *instruction,
 
     ProgramVariable receivingVar = ProgramVariable(store->getOperand(1));
 
-    for (auto Inst : workSet) {
-      if (valueToStore == Inst) {
-        if (CallInst *call = dyn_cast<CallInst>(Inst)) {
-          ProgramVariable callVar = ProgramVariable(call);
-          programPoint->addAlias(callVar, receivingVar);
+    if (CallInst *call = dyn_cast<CallInst>(valueToStore)) {
+      ProgramVariable callVar = ProgramVariable(call);
+      programPoint->addAlias(callVar, receivingVar);
+      return;
+    }
+
+    // check if two structs are being aliased
+    if (valueToStore->getType()->isPointerTy() &&
+        receivingValue->getType()->isPointerTy()) {
+      StructType *valueStruct =
+          dataflow::unwrapValuePointerToStruct(valueToStore);
+      StructType *receivingStruct =
+          dataflow::unwrapValuePointerToStruct(receivingValue);
+
+      if (valueStruct && receivingStruct) {
+        logout("two structs to alias " << *store);
+        int numFields = valueStruct->getNumElements();
+
+        // ASSUMPTION: both structs are referring to the same struct (or the
+        // structs have the same fields). note that i do not have strong reason
+        // to believe that they could be different, as the instruction is
+        // storing one struct into another.
```

Comment:
In general I don't think this is a safe assumption to make. C compilers generally will permit the programmer to assign a value of any struct type to any other struct type.

I think we need to write code here to conservatively handle the case where the structs have completely different layouts. In that case, we should generally try to just assume that anything could happen: that is, "go to top" and drop as much information as possible.

Reply:
This assumption is now dropped, and if they differ, we warn that aliasing information is lost.